pkg/ensign/broker/broker.go

```diff
@@ -0,0 +1,287 @@
+}
+
+// Run the broker; any fatal errors will be sent on the specified channel.
+func (b *Broker) Run(errc chan<- error) {
+	b.Lock()
+	defer b.Unlock()
+
+	// TODO: fetch list of topics
+
+	inQ := make(chan incoming, BufferSize)
+	outQ := make(chan *api.EventWrapper, BufferSize)
+	b.inQ = inQ
+
+	b.wg.Add(2)
+	go b.handleIncoming(inQ, outQ)
+	go b.handleOutgoing(outQ)
+}
+
+func (b *Broker) handleIncoming(inQ <-chan incoming, outQ chan<- *api.EventWrapper) {
+	defer b.wg.Done()
+	defer close(outQ)
+
+	seq := rlid.Sequence(0)
+	for incoming := range inQ {
+		// TODO: sequence RLIDs over topic offset instead of globally.
+		incoming.event.Id = seq.Next().Bytes()
+		incoming.event.Committed = timestamppb.Now()
+
+		// TODO: write event to disk
+		// TODO: consensus
+		// TODO: update topic metadata
+
+		// Send event on the outgoing queue
+		outQ <- incoming.event
+
+		// Send ack back to the publisher
+		b.pubmu.RLock()
+		if cb, ok := b.pubs[incoming.pubID]; ok {
+			res := PublishResult{
+				LocalID:   incoming.event.LocalId,
+				Committed: incoming.event.Committed.AsTime(),
+			}
+
+			// Non-blocking send so non-responding publishers don't hurt performance.
+			select {
+			case cb <- res:
+			default:
+			}
+		}
+		b.pubmu.RUnlock()
+
+		// Update metrics with number events
+		// TODO: update label values with topic name, publisher ID, node, and region
+		if o11y.Events != nil {
+			o11y.Events.WithLabelValues("unk", "unk").Inc()
+		}
+	}
+}
+
+func (b *Broker) handleOutgoing(outQ <-chan *api.EventWrapper) {
+	defer b.wg.Done()
+	for event := range outQ {
+		sends := 0
+		nsubs := 0
+
+		// Compute the topicID for the event
+		// TODO: how to handle topicID parsing errors?
+		topicID, _ := event.ParseTopicID()
+
+		b.submu.RLock()
+		for _, sub := range b.subs {
+			// Match the topic filter
+			if _, ok := sub.topics[topicID]; !ok {
+				continue
+			}
+
+			// Non-blocking send to prevent slow subscribers from interupting performance
+			nsubs++
+			select {
+			case sub.out <- event:
+				sends++
+			default:
+			}
+		}
+		b.submu.RUnlock()
+		log.Trace().Int("subs", sends).Bytes("id", event.Id).Int("dropped", nsubs-sends).Msg("event handled")
+	}
+}
+
+// Gracefully shutdown the broker. If a consensus or write operation is underway, then
+// shutdown blocks until it is concluded. The broker then stops handling incoming events
+// from publishers and closes all registered publishers and subscribers. This has the
+// effect of closing any open event stream handlers.
+func (b *Broker) Shutdown() error {
+	// Acquire a lock to close the the inQ channel and signal that we're no longer running.
+	b.Lock()
+
+	// If the broker is not running, ignore
+	if !b.isRunning() {
+		return nil
+	}
+
+	// Stop the internal go routines from handling any events
+	close(b.inQ)
+
+	// Unlock and wait for the incoming and outgoing go routines to stop processing.
+	b.Unlock()
+	b.wg.Wait()
+
+	b.Lock()
+	defer b.Unlock()
+
```

Comment:
How do we prevent publishers and subscribers from connecting while we're waiting?